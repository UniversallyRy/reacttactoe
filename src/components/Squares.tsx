import { calculateWinner } from "../util/gameAI";
import { SquareCSS } from "../view/styles";

export interface SquaresProps {
  state: {
    readonly squares: (string | null)[];
    readonly myTurn: boolean;
  };
  id: number;
  handleClick: ((id:number) => void);
}

export interface SquareCSSProps {
  player: string;
  onClick?: () => void;
  onTouchEnd?: (() => void);
  disabled: boolean;
}

 export const Square = ({ state, id, handleClick }:SquaresProps):JSX.Element => {
  let winner = calculateWinner(state.squares);
  let squares = [...state.squares];
  
  if (squares[id] === "O") {
    return (
      <SquareStyle disabled={true} player={"player"} />
    );
    // Does the same for AI
  } else if (squares[id] === "X") {
    return <SquareStyle disabled={true} player={"AI"} />;
  } else {
    // Default rendered squares when squares[i] has a null value
    return (
      <SquareStyle
        player={""}
        disabled={winner ? true : !state.myTurn}
        onTouchEnd={() => handleClick(id)}
        onClick={() => handleClick(id)}
      />
    );
  }
};

const SquareStyle = ({onClick, onTouchEnd, player, disabled }:SquareCSSProps) => {
  return (
    <SquareCSS
      player={player}
      style={{
        pointerEvents: disabled ? "none" : "auto",
      }}
      onTouchEnd={onTouchEnd}
      onClick={onClick}
    />
  );
};
