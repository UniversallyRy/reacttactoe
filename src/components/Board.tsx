import { useState } from "react";
import { Square } from "./Squares";
import { Overlay } from "./Overlay";
import { ResetButton } from "./ResetButton";
import { BoardWrap, GameTitle, Status, Reset, BoardRow, BoardBackground } from "../view/styles";
import { maximize } from "../util/gameAI";

export interface stateProps{
  readonly squares: (string | null)[] ;
  readonly myTurn: boolean;
}

export const Board = ():JSX.Element => {

  const [state, setState] = useState<stateProps>({
    squares: new Array(9).fill(null),
    myTurn: true,
  });

  const handleClick = (i:number) => {
    let squares:(string | null)[]  = [...state.squares];
    squares[i] = "O";
    // Sets current state of myTurn to false not disabling square clicks until AI chooses;
    setState((currentState) => {
      return { squares, myTurn: !currentState.myTurn, };
    });
    // setTimeout for AI's choice delay
    setTimeout(() => {
      let best:any = maximize(squares);
      squares[best[1]] = "X";
      setState((currentState) => {
        return { squares, myTurn: !currentState.myTurn };
      });
    }, 500);
  };

  const handleReset = () => {
    // Delay on reset to let above setTimout finish first
    setTimeout(() => {
      setState({ squares: Array(9).fill(null), myTurn: true });
    }, 300);
  };

  return (
    <BoardWrap>
      <GameTitle>React Tac Toe</GameTitle>
      <Status>{'You Go First'}</Status>
      <BoardBackground>
        <Overlay state={state} handleReset={handleReset}/>
        <Reset>
          <ResetButton state={state} handleReset={handleReset}/>
        </Reset>
        <BoardRow>
          <Square id={0} state={state} handleClick={handleClick}/>
          <Square id={1} state={state} handleClick={handleClick}/>
          <Square id={2} state={state} handleClick={handleClick}/>
        </BoardRow>
        <BoardRow>
          <Square id={3} state={state} handleClick={handleClick}/>
          <Square id={4} state={state} handleClick={handleClick}/>
          <Square id={5} state={state} handleClick={handleClick}/>
        </BoardRow>
        <BoardRow>
          <Square id={6} state={state} handleClick={handleClick}/>
          <Square id={7} state={state} handleClick={handleClick}/>
          <Square id={8} state={state} handleClick={handleClick}/>
        </BoardRow>
      </BoardBackground>
    </BoardWrap>
  );
};