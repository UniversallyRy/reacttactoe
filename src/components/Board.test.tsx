import ReactDOM from "react-dom";
import { render, screen, fireEvent } from '@testing-library/react';
import { Board } from "./Board";
import { Square } from "./Squares"

test("renders without crashing", async () => {
  const div = document.createElement("div");
  ReactDOM.render(<Board />, div);
});

test('renders reset button text', async () => {
  render(<Board />);
  const buttonElement = screen.getByText(/reset/i);
  expect(buttonElement).toBeInTheDocument();
});

test('renders square', async () => {
  const { container } = render(<Board />);
  const square = container.querySelector('.sc-iCoGMd.fXqpin')
  expect(square).toBeInTheDocument();
})

test('check default square color', async () => {
  const { container } = render(<Board />);
  const square = container.querySelector('.sc-iCoGMd.fXqpin')
  expect(square).toHaveStyleRule('background', '#cbb4d4')
  expect(square).toMatchSnapshot()
  expect(square).toBeInTheDocument();
});

test('check square presses', async () => {
  const state = {squares: [], myTurn:true}  
  const onClick = jest.fn();
  const { container } = render(<Square id={0} state={state} handleClick={onClick}/>);
  const square:any = container.querySelector('.sc-iCoGMd.fXqpin')
  expect(square).toBeInTheDocument();
  expect(onClick).toHaveBeenCalledTimes(0);
  fireEvent.click(square);
  expect(onClick).toHaveBeenCalledTimes(1);
});