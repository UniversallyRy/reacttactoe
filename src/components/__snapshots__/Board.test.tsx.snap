// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check default square color 1`] = `
.c0 {
  width: 20.833333333333336vw;
  min-height: 20.833333333333336vw;
  margin: 0.6em;
  background: #cbb4d4;
  color: #F00000;
  border: 1px dashed #cbb4d4;
  border-radius: 2px;
  text-shadow: 1px 0.6px #999;
  box-shadow: 0 0 0 3px #cbb4d4,1px 1px 6px 2px rgba(10,10,0,0.3);
  font-family: Fascinate,sans-serif;
  text-align: center;
}

.c0:focus {
  outline: none;
}

.c0:hover {
  background: #5f9ea0;
  opacity: 1;
  border: 1px dashed #5f9ea0;
  border-radius: 1px;
  box-shadow: 0 0 0 4px #5f9ea0,2px 1px 6px 4px rgba(10,10,0,0.5);
}

@media (min-width:768px) {

}

@media (min-width:1024px) {

}

@media (min-width:1024px) {

}

@media (min-width:768px) {

}

@media (min-width:1024px) {

}

@media (min-width:1024px) {

}

@media (min-width:768px) {
  .c0 {
    width: 17.36111111111111vw;
    min-height: 17.36111111111111vw;
    margin: 0.6em;
  }
}

@media (min-width:1024px) {
  .c0 {
    width: 8.333333333333332vw;
    min-height: 8.333333333333332vw;
  }
}

<button
  class="c0"
  style="pointer-events: auto;"
/>
`;
